src/fromager/dependency_graph.py

```diff
@@ -0,0 +1,272 @@
+import json
+import logging
+import pathlib
+import typing
+from enum import StrEnum
+
+from packaging.requirements import Requirement
+from packaging.utils import NormalizedName, canonicalize_name
+from packaging.version import Version
+
+logger = logging.getLogger(__name__)
+
+ROOT = ""
+
+
+class RequirementType(StrEnum):
+    INSTALL = "install"
+    TOP_LEVEL = "toplevel"
+    BUILD_SYSTEM = "build-system"
+    BUILD_BACKEND = "build-backend"
+    BUILD_SDIST = "build-sdist"
+
+
+class DependencyEdgeDict(typing.TypedDict):
+    req_type: str
+    req: str
+    key: str
+
+
+class DependencyNodeDict(typing.TypedDict):
+    download_url: str
+    canonicalized_name: str
+    version: str
+    edges: list[DependencyEdgeDict]
+
+
+class DependencyNode:
+    def __init__(
+        self,
+        req_name: NormalizedName | str,
+        version: Version | str,
+        download_url: str = "",
+    ) -> None:
+        self.canonicalized_name = canonicalize_name(req_name)
+        self.version = Version(str(version))
+
+        self.parents: list[DependencyEdge] = []
+        self.children: list[DependencyEdge] = []
+        self.key = f"{self.canonicalized_name}=={version}"
+        self.download_url = download_url
+
+    def add_child(
+        self,
+        child: "DependencyNode",
+        req: Requirement | str,
+        req_type: RequirementType | str,
+    ):
+        current_to_child_edge = DependencyEdge(
+            req=req, req_type=req_type, destination_node=child
+        )
+        self.children.append(current_to_child_edge)
+        child_to_current_edge = DependencyEdge(
+            req=req, req_type=req_type, destination_node=self
+        )
+        child.parents.append(child_to_current_edge)
```

Comment:
what would be the right way to do this then? Should I just store the node key instead of the entire node and then add a method to the graph that will fetch the node given the key?

Reply:
Storing only the key would work. A weakref is another option. I don't know that I'd bother with either here in a short-lived app, I just wanted to make sure you realized the cycle was there. If we have "fromager as a service" at some point, we'd want to resolve it. :-)